src/analyzing/AST_to_IL.ml

```diff
@@ -1236,55 +1240,82 @@ and mk_switch_break_label env tok =
   in
   (break_label, [ mk_s (Label break_label) ], switch_env)
 
+and implicit_return env eorig tok =
+  (* We always expect a value from an expression that is implicitly
+   * returned, so void is set to false here.
+   *)
+  let ss, e = expr_with_pre_stmts ~void:false env eorig in
+  let ret = mk_s (Return (tok, e)) in
+  ss @ [ ret ]
+
+and expr_stmt env (eorig : G.expr) tok : IL.stmt list =
+  (* optimize? pass context to expr when no need for return value? *)
+  let ss, e = expr_with_pre_stmts ~void:true env eorig in
+
+  (* ExprStmts don't care about results of an expression, so it is rightly
+   * setting void to true above. However, this means that function calls will
+   * return unit, and if we call mk_aux_var below, we will have something like
+   *   call f()
+   *   tmp = unit
+   * which interferes with implicit return analysis, because the  analysis walks
+   * backwards from the exit node to mark the first instr node it sees on each
+   * path.
+   *
+   * If we have
+   *   call f()
+   *   tmp = unit
+   * then `unit` will be marked as a returnning expression when we actually
+   * want to mark `f()`, so we must avoid creating `tmp = unit` following
+   * a function call that doesn't expect results.
+   *)
+  (match eorig with
+  | { e = G.Call _; _ } -> ()
+  | _else_ -> mk_aux_var env tok e |> ignore);
+
+  let ss' = pop_stmts env in
+  match ss @ ss' with
+  | [] ->
+      (* This case may happen when we have a function like
+       *
+       *   function some_function(some_var) {
+       *     some_var
```

Comment:
In this case, would not have the `mk_aux_var` above have created an assignment ?

Reply:
Calling `mk_aux_var` doesn't guarantee an assignment. When the expression passed is a Fetch, there's no assignment created. Previously I think I might have proposed some changes to `mk_aux_var`, but that was quite risky because `mk_aux_var` is used in many places.